### PR TITLE
Add Azure Blob Storage document loader

### LINKED ISSUE #438: [FEATURE] Add an Azure Blob Storage document loader
We could have an Azure Blob Storage document loader, like the existing AWS S3 loader at https://github.com/langchain4j/langchain4j/tree/main/document-loaders/langchain4j-document-loader-amazon-s3

I'm volunteering to code this!

### PR DESCRIPTION
**Pull request by jdubois**
Fix #438

Warning: this requires to upgrade Jackson to its latest version (`2.16.1`) so there might be an impact on other modules using it, like Elasticsearch.

### CODE REVIEW

File: document-loaders/langchain4j-document-loader-azure-storage-blob/pom.xml [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>dev.langchain4j</groupId>
+        <artifactId>langchain4j-parent</artifactId>
+        <version>0.26.0-SNAPSHOT</version>
+        <relativePath>../../langchain4j-parent/pom.xml</relativePath>
+    </parent>
+
+    <artifactId>langchain4j-document-loader-azure-storage-blob</artifactId>
```
Comment (langchain4j) [inline document-loaders/langchain4j-document-loader-azure-storage-blob/pom.xml:14]: Sorry for the nitpick, but why not `langchain4j-document-loader-azure-blob-storage`? 😄 as in "LangChain4j Azure Blob Storage document loader" and `AzureBlobStorageDocumentLoader`

File: document-loaders/langchain4j-document-loader-azure-storage-blob/src/main/java/dev/langchain4j/data/document/loader/azure/storage/blob/AzureBlobStorageDocumentLoader.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+                        documents.add(loadDocument(containerName, blob.getName(), parser)));
+
+        return documents;
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static class Builder {
```
Comment (langchain4j) [inline document-loaders/langchain4j-document-loader-azure-storage-blob/src/main/java/dev/langchain4j/data/document/loader/azure/storage/blob/AzureBlobStorageDocumentLoader.java]: This builder has only one field, seems redundant

File: document-loaders/langchain4j-document-loader-azure-storage-blob/src/main/java/dev/langchain4j/data/document/source/azure/storage/blob/AzureBlobStorageSource.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+    private final InputStream inputStream;
+    private final String accountName;
+    private final String containerName;
+    private final String blobName;
+    private final BlobProperties properties;
+
+    public AzureBlobStorageSource(InputStream inputStream, String containerName, String accountName, String blobNAme, BlobProperties properties) {
+        this.inputStream = ensureNotNull(inputStream, "inputStream");
+        this.accountName = ensureNotBlank(containerName, "accountName");
```
Comment (langchain4j) [inline document-loaders/langchain4j-document-loader-azure-storage-blob/src/main/java/dev/langchain4j/data/document/source/azure/storage/blob/AzureBlobStorageSource.java]: containerName -> accountName

File: document-loaders/langchain4j-document-loader-azure-storage-blob/src/main/java/dev/langchain4j/data/document/source/azure/storage/blob/AzureBlobStorageSource.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+    public static final String SOURCE = "source";
+
+    private final InputStream inputStream;
+    private final String accountName;
+    private final String containerName;
+    private final String blobName;
+    private final BlobProperties properties;
+
+    public AzureBlobStorageSource(InputStream inputStream, String containerName, String accountName, String blobNAme, BlobProperties properties) {
```
Comment (langchain4j) [inline document-loaders/langchain4j-document-loader-azure-storage-blob/src/main/java/dev/langchain4j/data/document/source/azure/storage/blob/AzureBlobStorageSource.java]: minor: blobNAme -> blobName

File: document-loaders/langchain4j-document-loader-azure-storage-blob/src/main/java/dev/langchain4j/data/document/source/azure/storage/blob/AzureBlobStorageSource.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    @Override
+    public InputStream inputStream() {
+        return inputStream;
+    }
+
+    @Override
+    public Metadata metadata() {
+        Metadata metadata = new Metadata();
+        metadata.add(SOURCE, format("https://%s.blob.core.windows.net/%s/%s", accountName, containerName, blobName));
+        metadata.add("Creation-Time", properties.getCreationTime());
```
Comment (langchain4j) [inline document-loaders/langchain4j-document-loader-azure-storage-blob/src/main/java/dev/langchain4j/data/document/source/azure/storage/blob/AzureBlobStorageSource.java]: WDYT about prepending metadata keys with "azure_blob_" suffix and converting into snake_case for uniformity with other loaders?

File: document-loaders/langchain4j-document-loader-azure-storage-blob/pom.xml [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+        <dependency>
+            <groupId>org.assertj</groupId>
+            <artifactId>assertj-core</artifactId>
+            <scope>test</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>org.testcontainers</groupId>
+            <artifactId>localstack</artifactId>
```
Comment (langchain4j) [inline document-loaders/langchain4j-document-loader-azure-storage-blob/pom.xml]: seems not used

**Review by ghost — CHANGES_REQUESTED** [review]
@jdubois thanks a lot! I have left a few minor comments

**Comment by langchain4j** [comment]
@jdubois BTW why (specific version of) jackson is required?

File: document-loaders/langchain4j-document-loader-azure-storage-blob/pom.xml (hunk 1 above)
Comment (jdubois) [reply document-loaders/langchain4j-document-loader-azure-storage-blob/pom.xml:14]: Ah, that's a naming problem with Azure Blob Storage.
The library's name (as in Maven) is `azure-storage-blob` which makes sense, as there are some other storage mechanisms (like Table Storage and File Storage). And that's also what is being used in URLs like https://learn.microsoft.com/en-us/azure/storage/blobs/
Then, when you say the name in English, people will call it "Azure Blob Storage".

So in the Azure SDK you'll have `com.azure.storage.blob.implementation.AzureBlobStorageImpl`, basically with the package name using "storage-blob" and the Class name using "BlobStorage".

This is what I'm trying to follow here.

File: document-loaders/langchain4j-document-loader-azure-storage-blob/src/main/java/dev/langchain4j/data/document/loader/azure/storage/blob/AzureBlobStorageDocumentLoader.java (hunk 2 above)
Comment (jdubois) [reply document-loaders/langchain4j-document-loader-azure-storage-blob/src/main/java/dev/langchain4j/data/document/loader/azure/storage/blob/AzureBlobStorageDocumentLoader.java]: I was trying to follow the rest of the API that uses Builder pattern, is it fine for you if I do 2 constructors, an empty one (which will create the `blobServiceClient`) and one that takes the `blobServiceClient` as parameter?

Comment (langchain4j) [reply document-loaders/langchain4j-document-loader-azure-storage-blob/src/main/java/dev/langchain4j/data/document/loader/azure/storage/blob/AzureBlobStorageDocumentLoader.java]: @jdubois sounds good!

Comment (jdubois) [reply document-loaders/langchain4j-document-loader-azure-storage-blob/src/main/java/dev/langchain4j/data/document/loader/azure/storage/blob/AzureBlobStorageDocumentLoader.java]: In fact I'm not keeping the empty constructor, it makes no sense as the endpoint and the credentials are always required

File: document-loaders/langchain4j-document-loader-azure-storage-blob/src/main/java/dev/langchain4j/data/document/source/azure/storage/blob/AzureBlobStorageSource.java (hunk 5 above)
Comment (jdubois) [reply document-loaders/langchain4j-document-loader-azure-storage-blob/src/main/java/dev/langchain4j/data/document/source/azure/storage/blob/AzureBlobStorageSource.java]: I'm using `azure_storage_blob_`

**Comment by jdubois** [comment]
@langchain4j 
- For Jackson, I'm not even sure it's being used here, but the Azure SDK won't start if you're not having a recent version. There's a check at the beginning so they are sure you use a compatible version of Jackson.
- Otherwise I've fixed all the comments, and did the integration tests, we should be all OK

**Review by ghost — APPROVED** [review]
@jdubois looks great, thank you!

**Comment by langchain4j** [comment]
@jdubois could you please check the tests? I am getting this error:

```
com.azure.storage.blob.models.BlobStorageException: Status code 400, "<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<Error>
  <Code>InvalidHeaderValue</Code>
  <Message>The API version 2024-11-04 is not supported by Azurite. Please upgrade Azurite to latest version and retry. If you are using Azurite in Visual Studio, please check you have installed latest Visual Studio patch. Azurite command line parameter "--skipApiVersionCheck" or Visual Studio Code configuration "Skip Api Version Check" can skip this error. 
RequestId:84805810-973e-486d-b8de-e21c3cfc09b1
Time:2024-11-13T12:50:16.382Z</Message>
</Error>"

	at com.azure.storage.blob.implementation.util.ModelHelper.mapToBlobStorageException(ModelHelper.java:483)
	at reactor.core.publisher.Mono.lambda$onErrorMap$30(Mono.java:3797)
	at reactor.core.publisher.Mono.lambda$onErrorResume$32(Mono.java:3887)
	at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onError(FluxOnErrorResume.java:94)
	at reactor.core.publisher.MonoFlatMap$FlatMapMain.onError(MonoFlatMap.java:172)
	at reactor.core.publisher.MonoFlatMap$FlatMapMain.secondError(MonoFlatMap.java:192)
	at reactor.core.publisher.MonoFlatMap$FlatMapInner.onError(MonoFlatMap.java:259)
	at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:142)
	at reactor.core.publisher.FluxSwitchIfEmpty$SwitchIfEmptySubscriber.onNext(FluxSwitchIfEmpty.java:74)
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onNext(FluxMapFuseable.java:129)
	at reactor.core.publisher.Operators$ScalarSubscription.request(Operators.java:2400)
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.request(FluxMapFuseable.java:171)
	at reactor.core.publisher.Operators$MultiSubscriptionSubscriber.set(Operators.java:2196)
	at reactor.core.publisher.Operators$MultiSubscriptionSubscriber.onSubscribe(Operators.java:2070)
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onSubscribe(FluxMapFuseable.java:96)
	at reactor.core.publisher.MonoJust.subscribe(MonoJust.java:55)
	at reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:64)
	at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:157)
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onNext(FluxMapFuseable.java:129)
	at reactor.core.publisher.FluxHide$SuppressFuseableSubscriber.onNext(FluxHide.java:137)
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onNext(FluxMapFuseable.java:129)
	at reactor.core.publisher.FluxHide$SuppressFuseableSubscriber.onNext(FluxHide.java:137)
	at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onNext(FluxOnErrorResume.java:79)
	at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1839)
	at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:151)
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onNext(FluxMapFuseable.java:129)
	at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1839)
	at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:151)
	at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onNext(FluxOnErrorResume.java:79)
	at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1839)
	at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:151)
	at reactor.core.publisher.Operators$MonoInnerProducerBase.complete(Operators.java:2666)
	at reactor.core.publisher.MonoSingle$SingleSubscriber.onComplete(MonoSingle.java:180)
	at reactor.core.publisher.MonoFlatMapMany$FlatMapManyInner.onComplete(MonoFlatMapMany.java:260)
	at reactor.core.publisher.FluxContextWrite$ContextWriteSubscriber.onComplete(FluxContextWrite.java:126)
	at reactor.core.publisher.MonoUsing$MonoUsingSubscriber.onNext(MonoUsing.java:232)
	at reactor.core.publisher.FluxMap$MapSubscriber.onNext(FluxMap.java:122)
	at reactor.core.publisher.FluxSwitchIfEmpty$SwitchIfEmptySubscriber.onNext(FluxSwitchIfEmpty.java:74)
	at reactor.core.publisher.FluxHandle$HandleSubscriber.onNext(FluxHandle.java:126)
	at reactor.core.publisher.FluxMap$MapConditionalSubscriber.onNext(FluxMap.java:224)
	at reactor.core.publisher.FluxDoFinally$DoFinallySubscriber.onNext(FluxDoFinally.java:113)
	at reactor.core.publisher.FluxHandleFuseable$HandleFuseableSubscriber.onNext(FluxHandleFuseable.java:191)
	at reactor.core.publisher.FluxContextWrite$ContextWriteSubscriber.onNext(FluxContextWrite.java:107)
	at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1839)
	at reactor.core.publisher.MonoCollectList$MonoCollectListSubscriber.onComplete(MonoCollectList.java:129)
	at reactor.core.publisher.FluxPeek$PeekSubscriber.onComplete(FluxPeek.java:260)
	at reactor.core.publisher.FluxMap$MapSubscriber.onComplete(FluxMap.java:144)
	at reactor.netty.channel.FluxReceive.onInboundComplete(FluxReceive.java:415)
	at reactor.netty.channel.ChannelOperations.onInboundComplete(ChannelOperations.java:439)
	at reactor.netty.channel.ChannelOperations.terminate(ChannelOperations.java:493)
	at reactor.netty.http.client.HttpClientOperations.onInboundNext(HttpClientOperations.java:789)
	at reactor.netty.channel.ChannelOperationsHandler.channelRead(ChannelOperationsHandler.java:114)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at com.azure.core.http.netty.implementation.AzureSdkHandler.channelRead(AzureSdkHandler.java:224)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:318)
	at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1407)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:918)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:994)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:1583)
	Suppressed: java.lang.Exception: #block terminated with an error
		at reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:100)
		at reactor.core.publisher.Mono.block(Mono.java:1742)
		at com.azure.storage.common.implementation.StorageImplUtils.blockWithOptionalTimeout(StorageImplUtils.java:146)
		at com.azure.storage.blob.BlobContainerClient.createIfNotExistsWithResponse(BlobContainerClient.java:375)
		at com.azure.storage.blob.BlobServiceClient.createBlobContainerIfNotExistsWithResponse(BlobServiceClient.java:211)
		at com.azure.storage.blob.BlobServiceClient.createBlobContainerIfNotExists(BlobServiceClient.java:173)
		at dev.langchain4j.data.document.loader.azure.storage.blob.LocalAzureBlobStorageDocumentLoaderIT.beforeAll(LocalAzureBlobStorageDocumentLoaderIT.java:52)
		at java.base/java.lang.reflect.Method.invoke(Method.java:580)
		at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	Suppressed: com.azure.storage.blob.models.BlobStorageException: Status code 400, "<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<Error>
  <Code>InvalidHeaderValue</Code>
  <Message>The API version 2024-11-04 is not supported by Azurite. Please upgrade Azurite to latest version and retry. If you are using Azurite in Visual Studio, please check you have installed latest Visual Studio patch. Azurite command line parameter "--skipApiVersionCheck" or Visual Studio Code configuration "Skip Api Version Check" can skip this error. 
RequestId:4475d910-5b2d-41fa-9382-a50252a1be02
Time:2024-11-13T12:50:16.475Z</Message>
</Error>"
		... 79 more
		Suppressed: java.lang.Exception: #block terminated with an error
			at reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:100)
			at reactor.core.publisher.Mono.block(Mono.java:1742)
			at com.azure.storage.common.implementation.StorageImplUtils.blockWithOptionalTimeout(StorageImplUtils.java:146)
			at com.azure.storage.blob.BlobContainerClient.deleteWithResponse(BlobContainerClient.java:433)
			at com.azure.storage.blob.BlobContainerClient.delete(BlobContainerClient.java:401)
			at dev.langchain4j.data.document.loader.azure.storage.blob.LocalAzureBlobStorageDocumentLoaderIT.afterAll(LocalAzureBlobStorageDocumentLoaderIT.java:86)
			at java.base/java.lang.reflect.Method.invoke(Method.java:580)
			at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
			at java.base/java.util.Collections$UnmodifiableCollection.forEach(Collections.java:1116)
			at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
```

**Comment by jdubois** [comment]
@langchain4j which Azurite version are you using? According to https://github.com/Azure/Azurite/blob/49a26219f99650cff891fc64c52ecaab5ee9c464/ChangeLog.md?plain=1#L25 they should support Api version 2024-11-04 in Azurite 3.32.0

**Comment by langchain4j** [comment]
@jdubois I am using the one that is configured in `LocalAzureBlobStorageDocumentLoaderIT`: `azurite:latest`.
Is this test green for you?

**Comment by jdubois** [comment]
Oh I had no idea @eddumelendez added a test using TestContainers + Azurite, that's super cool!!
It currently works on my machine as of today, then it's using "azurite:latest" so my guess is that there has been a version supporting that API that just got out. That's both the problem and the solution of using the "latest" tag.
I just had a similar issue in JHipster Online at https://github.com/jhipster/jhipster-online/pull/451 as we got to a newer version automatically and that broke our build.... and @eddumelendez is also in the thread discussing the solution, so I guess he knows all this a ton better than me. WDYT?

**Comment by eddumelendez** [comment]
works for me as well. What we can do is adding `.withCommand("--skipApiVersionCheck")` and use a specific version such as `3.33.0`. Azurite images available are [here](https://mcr.microsoft.com/v2/azure-storage/azurite/tags/list)

**Comment by langchain4j** [comment]
@jdubois @eddumelendez thanks for clarifications! It was still failing for me, so I removed azurite image from docker and now the test is green. Any idea why does this happen?

**Comment by jdubois** [comment]
You probably had an older version in the cache ("latest" doesn't specify a version), and so you updated it that way, which fixed the issue.

**Comment by langchain4j** [comment]
Interesting, I thought that "latest" means "check if there is a newer version before run and download it". Apparently it is now how it works. Thanks for clarification!

**Pull request merged**